File: Week-Wise-Assignments/Assignment 9/Compressed Solutions/Python Programs/1_binary_search_tree.py
class Node:
    def __init__(self, data):
        self.data, self.left, self.right = data, None, None

class BST:
    def __init__(self):
        self.root = None
    
    def insert(self, data):
        self.root = self._insert(self.root, data)
    
    def _insert(self, root, data):
        if not root: return Node(data)
        if data < root.data: root.left = self._insert(root.left, data)
        elif data > root.data: root.right = self._insert(root.right, data)
        return root
    
    def delete(self, data):
        self.root = self._delete(self.root, data)
    
    def _delete(self, root, data):
        if not root: return root
        if data < root.data: root.left = self._delete(root.left, data)
        elif data > root.data: root.right = self._delete(root.right, data)
        else:
            if not root.left: return root.right
            if not root.right: return root.left
            min_node = self._min(root.right)
            root.data = min_node.data
            root.right = self._delete(root.right, min_node.data)
        return root
    
    def search(self, key):
        return self._search(self.root, key)
    
    def _search(self, root, key):
        if not root or root.data == key: return root
        return self._search(root.left, key) if key < root.data else self._search(root.right, key)
    
    def inorder(self, root=None):
        if root is None: root = self.root
        return (self.inorder(root.left) + [root.data] + self.inorder(root.right)) if root else []
    
    def preorder(self, root=None):
        if root is None: root = self.root
        return ([root.data] + self.preorder(root.left) + self.preorder(root.right)) if root else []
    
    def postorder(self, root=None):
        if root is None: root = self.root
        return (self.postorder(root.left) + self.postorder(root.right) + [root.data]) if root else []
    
    def count(self, root=None):
        if root is None: root = self.root
        return 1 + self.count(root.left) + self.count(root.right) if root else 0
    
    def height(self, root=None):
        if root is None: root = self.root
        return 1 + max(self.height(root.left), self.height(root.right)) if root else -1
    
    def leafs(self, root=None):
        if root is None: root = self.root
        if not root: return 0
        if not root.left and not root.right: return 1
        return self.leafs(root.left) + self.leafs(root.right)
    
    def internal(self, root=None):
        if root is None: root = self.root
        if not root or (not root.left and not root.right): return 0
        return 1 + self.internal(root.left) + self.internal(root.right)
    
    def _min(self, root):
        while root and root.left: root = root.left
        return root
    
    def _max(self, root):
        while root and root.right: root = root.right
        return root
    
    def min_val(self):
        node = self._min(self.root)
        return node.data if node else None
    
    def max_val(self):
        node = self._max(self.root)
        return node.data if node else None
    
    def is_complete(self, root=None, i=0, n=None):
        if root is None: root, n = self.root, self.count()
        if not root: return True
        if i >= n: return False
        return self.is_complete(root.left, 2*i+1, n) and self.is_complete(root.right, 2*i+2, n)
    
    def balance(self, root=None):
        if root is None: root = self.root
        return self.height(root.left) - self.height(root.right) if root else 0
    
    def null_count(self, root=None):
        if root is None: root = self.root
        return self.null_count(root.left) + self.null_count(root.right) if root else 1

bst = BST()
while True:
    print("\n1.Create 2.Insert 3.Delete 4.Search 5.Inorder 6.Preorder 7.Postorder 8.Count 9.Height 10.Leafs 11.Internal 12.Min 13.Max 14.Complete 15.Balance 16.NullCount 0.Exit")
    ch = int(input())
    if ch == 1:
        n = int(input("Nodes: "))
        for _ in range(n):
            bst.insert(int(input("Data: ")))
    elif ch == 2: bst.insert(int(input("Data: ")))
    elif ch == 3: bst.delete(int(input("Data: ")))
    elif ch == 4: print("Found" if bst.search(int(input("Key: "))) else "Not found")
    elif ch == 5: print("Inorder:", bst.inorder())
    elif ch == 6: print("Preorder:", bst.preorder())
    elif ch == 7: print("Postorder:", bst.postorder())
    elif ch == 8: print("Nodes:", bst.count())
    elif ch == 9: print("Height:", bst.height())
    elif ch == 10: print("Leafs:", bst.leafs())
    elif ch == 11: print("Internal:", bst.internal())
    elif ch == 12: print("Min:", bst.min_val())
    elif ch == 13: print("Max:", bst.max_val())
    elif ch == 14: print("Complete:", "Yes" if bst.is_complete() else "No")
    elif ch == 15: print("Balance:", bst.balance())
    elif ch == 16: print("NULL:", bst.null_count())
    elif ch == 0: break

File: Week-Wise-Assignments/Assignment 9/Compressed Solutions/C Programs/1_binary_search_tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *left, *right;
} Node;

Node* create(int data) {
    Node* n = malloc(sizeof(Node));
    n->data = data; n->left = n->right = NULL;
    return n;
}

Node* insert(Node* root, int data) {
    if(!root) return create(data);
    if(data < root->data) root->left = insert(root->left, data);
    else if(data > root->data) root->right = insert(root->right, data);
    return root;
}

Node* findMin(Node* root) {
    while(root && root->left) root = root->left;
    return root;
}

Node* delete(Node* root, int data) {
    if(!root) return root;
    if(data < root->data) root->left = delete(root->left, data);
    else if(data > root->data) root->right = delete(root->right, data);
    else {
        if(!root->left) { Node* temp = root->right; free(root); return temp; }
        if(!root->right) { Node* temp = root->left; free(root); return temp; }
        Node* temp = findMin(root->right);
        root->data = temp->data;
        root->right = delete(root->right, temp->data);
    }
    return root;
}

void inorder(Node* root) {
    if(root) { inorder(root->left); printf("%d ", root->data); inorder(root->right); }
}

void preorder(Node* root) {
    if(root) { printf("%d ", root->data); preorder(root->left); preorder(root->right); }
}

void postorder(Node* root) {
    if(root) { postorder(root->left); postorder(root->right); printf("%d ", root->data); }
}

int count(Node* root) {
    return root ? 1 + count(root->left) + count(root->right) : 0;
}

int height(Node* root) {
    if(!root) return -1;
    int l = height(root->left), r = height(root->right);
    return 1 + (l > r ? l : r);
}

int leafs(Node* root) {
    if(!root) return 0;
    if(!root->left && !root->right) return 1;
    return leafs(root->left) + leafs(root->right);
}

int internal(Node* root) {
    if(!root || (!root->left && !root->right)) return 0;
    return 1 + internal(root->left) + internal(root->right);
}

Node* search(Node* root, int key) {
    if(!root || root->data == key) return root;
    return key < root->data ? search(root->left, key) : search(root->right, key);
}

int isComplete(Node* root, int i, int n) {
    if(!root) return 1;
    if(i >= n) return 0;
    return isComplete(root->left, 2*i+1, n) && isComplete(root->right, 2*i+2, n);
}

int balance(Node* root) {
    return root ? height(root->left) - height(root->right) : 0;
}

Node* findMax(Node* root) {
    while(root && root->right) root = root->right;
    return root;
}

int nullCount(Node* root) {
    return root ? nullCount(root->left) + nullCount(root->right) : 1;
}

int main() {
    Node* root = NULL;
    int ch, data, n;
    while(1) {
        printf("\n1.Create 2.Insert 3.Delete 4.Search 5.Inorder 6.Preorder 7.Postorder 8.Count 9.Height 10.Leafs 11.Internal 12.Min 13.Max 14.Complete 15.Balance 16.NullCount 0.Exit\n");
        scanf("%d", &ch);
        switch(ch) {
            case 1:
                printf("Nodes: "); scanf("%d", &n);
                for(int i = 0; i < n; i++) {
                    printf("Data: "); scanf("%d", &data);
                    root = insert(root, data);
                }
                break;
            case 2: printf("Data: "); scanf("%d", &data); root = insert(root, data); break;
            case 3: printf("Data: "); scanf("%d", &data); root = delete(root, data); break;
            case 4: printf("Key: "); scanf("%d", &data); printf("%s\n", search(root, data) ? "Found" : "Not found"); break;
            case 5: printf("Inorder: "); inorder(root); printf("\n"); break;
            case 6: printf("Preorder: "); preorder(root); printf("\n"); break;
            case 7: printf("Postorder: "); postorder(root); printf("\n"); break;
            case 8: printf("Nodes: %d\n", count(root)); break;
            case 9: printf("Height: %d\n", height(root)); break;
            case 10: printf("Leafs: %d\n", leafs(root)); break;
            case 11: printf("Internal: %d\n", internal(root)); break;
            case 12: { Node* min = findMin(root); printf("Min: %s\n", min ? (printf("%d", min->data), "") : "Empty"); } break;
            case 13: { Node* max = findMax(root); printf("Max: %s\n", max ? (printf("%d", max->data), "") : "Empty"); } break;
            case 14: printf("Complete: %s\n", isComplete(root, 0, count(root)) ? "Yes" : "No"); break;
            case 15: printf("Balance: %d\n", balance(root)); break;
            case 16: printf("NULL: %d\n", nullCount(root)); break;
            case 0: return 0;
        }
    }
}